### PR TITLE
Simple Berkeley Socket APIs

### PR DESCRIPTION
**Pull request by lum1n0us**
Reference:
[Networking API design] (https://github.com/WebAssembly/WASI/issues/370)

### CODE REVIEW

**Comment by lum1n0us** [comment]
It is a simple socket example and related APIs implementation. The goal is to make it easy to compile a C socket program. 

The first version ignores the capability security model.

File: core/iwasm/libraries/libc-wasi/libc_wasi_wrapper.c [hunk 1]
```diff
@@ -986,6 +995,184 @@ wasi_random_get(wasm_exec_env_t exec_env, void *buf, uint32 buf_len)
...
+    wasm_module_inst_t module_inst = get_module_inst(exec_env);
+    wasi_ctx_t wasi_ctx = get_wasi_ctx(module_inst);
+    struct fd_table *curfds = NULL;
+    struct addr_pool *addr_pool = NULL;
+
+    if (!wasi_ctx)
+        return __WASI_EACCES;
+
+    curfds = wasi_ctx_get_curfds(module_inst, wasi_ctx);
+    addr_pool = wasi_ctx_get_addr_pool(module_inst, wasi_ctx);
```
Comment (wenyongh) [inline core/iwasm/libraries/libc-wasi/libc_wasi_wrapper.c:1064]: Shall we check whether addr_pool is NULL or not?

File: core/iwasm/libraries/libc-wasi/libc_wasi_wrapper.c [hunk 2]
```diff
@@ -986,6 +995,184 @@ wasi_random_get(wasm_exec_env_t exec_env, void *buf, uint32 buf_len)
...
+
+    return wasi_ssp_sock_bind(curfds, addr_pool, fd, addr);
```
Comment (wenyongh) [inline core/iwasm/libraries/libc-wasi/libc_wasi_wrapper.c:1066]: What is the implementation of wasi_ssp_sock_bind? Cannot find it this patch?

File: core/iwasm/libraries/libc-wasi/libc_wasi_wrapper.c [hunk 3]
```diff
@@ -994,51 +1181,38 @@ wasi_sock_recv(wasm_exec_env_t exec_env, wasi_fd_t sock, iovec_app_t *ri_data,
...
+        err = wasmtime_ssp_sock_recv(curfds, sock, buf, ri_data->buf_len);
+        if (err != __WASI_ESUCCESS) {
+            return err;
+        }
+    }
 
-fail:
-    wasm_runtime_free(iovec_begin);
-    return err;
+    *(uint32 *)ro_datalen_app = ri_data_len;
```
Comment (wenyongh) [inline core/iwasm/libraries/libc-wasi/libc_wasi_wrapper.c:1213]: ri_data_len seems to be the packet count, is it right to set it (but not total ri_data->buf_len) to ro_datalen_app?

File: core/iwasm/libraries/libc-wasi/libc_wasi_wrapper.c [hunk 4]
```diff
@@ -1049,50 +1223,36 @@ wasi_sock_send(wasm_exec_env_t exec_env, wasi_fd_t sock,
...
+        err = wasmtime_ssp_sock_send(curfds, sock, buf, si_data->buf_len);
+        if (err != __WASI_ESUCCESS) {
+            return err;
+        }
+    }
 
-fail:
-    wasm_runtime_free(ciovec_begin);
-    return err;
+    *so_datalen_app = (uint32)si_data_len;
```
Comment (wenyongh) [inline core/iwasm/libraries/libc-wasi/libc_wasi_wrapper.c:1254]: Same as above

File: core/iwasm/libraries/libc-wasi/sandboxed-system-primitives/include/wasmtime_ssp.h [hunk 5]
```diff
@@ -536,6 +549,55 @@ _Static_assert(_Alignof(__wasi_subscription_t) == 8, "witx calculated align");
...
+    __wasi_ip_port_t port;
+} __wasi_addr_ip6_port_t;
+
+typedef struct __wasi_addr_t {
+    __wasi_addr_type_t kind;
+    union {
+        __wasi_addr_ip4_port_t ip4;
+        __wasi_addr_ip6_port_t ip6;
+    } addr;
+} __wasi_addr_t;
```
Comment (wenyongh) [inline core/iwasm/libraries/libc-wasi/sandboxed-system-primitives/include/wasmtime_ssp.h:597]: Can the layout of these structures be same in 32-bit and 64-bit system? Had we tested it in x86-32? If not the same, should add the related padding bytes in the structure.

File: core/iwasm/libraries/lib-socket/inc/wasi_socket_ext.h [hunk 6]
```diff
@@ -0,0 +1,410 @@
...
+int32_t
+__imported_wasi_snapshot_preview1_sock_accept(int32_t arg0, int32_t arg1)
+    __attribute__((__import_module__("wasi_snapshot_preview1"),
+                   __import_name__("sock_accept")));
+
+static inline __wasi_errno_t
+__wasi_sock_accept(__wasi_fd_t fd, __wasi_fd_t *fd_new)
+{
+    return (__wasi_errno_t)__imported_wasi_snapshot_preview1_sock_accept(
+        (int32_t)fd, (int32_t)fd_new);
```
[the remark this answers is not in the document]
Comment (lum1n0us) [reply core/iwasm/libraries/lib-socket/inc/wasi_socket_ext.h:100]: This piece of code is a patch for [__wasilibc_real.c](https://github.com/WebAssembly/wasi-libc/blob/main/libc-bottom-half/sources/__wasilibc_real.c).  Using `(int32_t)` to represent a memory address is a routine. If considering that there are only two integer types in the WASM world and WASM is still in WASM-32, `int32_t` casting seems reasonable.

File: core/iwasm/libraries/lib-socket/src/wasi/wasi_socket_ext.c [hunk 7]
```diff
@@ -0,0 +1,181 @@
...
+    if (IPv4 == wasi_addr.kind) {
+        struct sockaddr_in sock_addr_in = { 0 };
+
+        sock_addr_in.sin_family = AF_INET;
+        sock_addr_in.sin_addr.s_addr = (wasi_addr.addr.ip4.addr.n3 << 24)
+                                       | (wasi_addr.addr.ip4.addr.n2 << 16)
+                                       | (wasi_addr.addr.ip4.addr.n1 << 8)
+                                       | wasi_addr.addr.ip4.addr.n0;
+        sock_addr_in.sin_port = htons(wasi_addr.addr.ip4.port);
+        memcpy(sock_addr, &sock_addr_in, sizeof(sock_addr_in));
```
[the remark this answers is not in the document]
Comment (lum1n0us) [reply core/iwasm/libraries/lib-socket/src/wasi/wasi_socket_ext.c:77]: It is a part of the *wasi-libc* and can not use any `bh_xxx` APIs

File: core/iwasm/libraries/libc-wasi/libc_wasi_wrapper.c (hunk 2 above)
Comment (lum1n0us) [reply core/iwasm/libraries/libc-wasi/libc_wasi_wrapper.c:1066]: In `core/iwasm/libraries/libc-wasi/sandboxed-system-primitives/src/posix.c`

File: core/iwasm/libraries/libc-wasi/libc_wasi_wrapper.c (hunk 1 above)
Comment (lum1n0us) [reply core/iwasm/libraries/libc-wasi/libc_wasi_wrapper.c:1064]: kindly not sure about that. If `wasm_runtime_init_wasi()` is always executed successfully and earlier, `addr_pool` should like other key items, like `curfds`.

File: core/iwasm/libraries/libc-wasi/libc_wasi_wrapper.c (hunk 3 above)
Comment (lum1n0us) [reply core/iwasm/libraries/libc-wasi/libc_wasi_wrapper.c:1213]: Right. should make it right.

File: core/iwasm/libraries/libc-wasi/libc_wasi_wrapper.c (hunk 2 above)
Comment (wenyongh) [reply core/iwasm/libraries/libc-wasi/libc_wasi_wrapper.c:1066]: OK

File: core/iwasm/libraries/libc-wasi/libc_wasi_wrapper.c (hunk 3 above)
Comment (wenyongh) [reply core/iwasm/libraries/libc-wasi/libc_wasi_wrapper.c:1213]: OK

File: core/iwasm/libraries/libc-wasi/sandboxed-system-primitives/src/posix.c [hunk 8]
```diff
@@ -2716,113 +2717,286 @@ wasmtime_ssp_random_get(void *buf, size_t nbyte)
...
+    ret = os_socket_accept(fd_number(fo), &new_sock, NULL, NULL);
     fd_object_release(fo);
-    if (datalen < 0) {
+    if (ret == BHT_ERROR)
         return convert_errno(errno);
+
+    error = fd_determine_type_rights(new_sock, &wasi_type, &max_base,
+                                     &max_inheriting);
+    if (error != __WASI_ESUCCESS) {
+        close(ret);
```
Comment (wenyongh) [inline core/iwasm/libraries/libc-wasi/sandboxed-system-primitives/src/posix.c:2744]: Should be os_socket_close but not close? As it is opened be os_socket_accept

File: core/iwasm/libraries/libc-wasi/sandboxed-system-primitives/src/posix.c [hunk 9]
```diff
@@ -2716,113 +2717,286 @@ wasmtime_ssp_random_get(void *buf, size_t nbyte)
...
-    // Convert msghdr to output.
-    *ro_datalen = (size_t)datalen;
-    *ro_flags = 0;
-    if ((hdr.msg_flags & MSG_TRUNC) != 0)
-        *ro_flags |= __WASI_SOCK_RECV_DATA_TRUNCATED;
-    return 0;
+    error = fd_table_insert_fd(curfds, new_sock, wasi_type, max_base,
+                               max_inheriting, fd_new);
+    if (error != __WASI_ESUCCESS) {
+        close(ret);
```
Comment (wenyongh) [inline core/iwasm/libraries/libc-wasi/sandboxed-system-primitives/src/posix.c:2751]: Sames above

File: core/iwasm/libraries/libc-wasi/sandboxed-system-primitives/src/posix.c [hunk 10]
```diff
@@ -2716,113 +2717,286 @@ wasmtime_ssp_random_get(void *buf, size_t nbyte)
...
+
+    ret = os_socket_create(&sock, tcp_or_udp);
+    if (ret == BHT_ERROR) {
+        return convert_errno(errno);
+    }
+
+    error =
+        fd_determine_type_rights(sock, &wasi_type, &max_base, &max_inheriting);
+    if (error != __WASI_ESUCCESS) {
+        close(ret);
```
Comment (wenyongh) [inline core/iwasm/libraries/libc-wasi/sandboxed-system-primitives/src/posix.c:2913]: Same as above

File: core/iwasm/libraries/libc-wasi/sandboxed-system-primitives/src/posix.c [hunk 11]
```diff
@@ -2716,113 +2717,286 @@ wasmtime_ssp_random_get(void *buf, size_t nbyte)
...
     }
     else {
-        *so_datalen = (size_t)len;
+        assert(wasi_type == __WASI_FILETYPE_SOCKET_STREAM);
     }
 
-out:
-    return error;
+    // TODO: base rights and inheriting rights ?
+    return fd_table_insert_fd(curfds, sock, wasi_type, max_base, max_inheriting,
```
Comment (wenyongh) [inline core/iwasm/libraries/libc-wasi/sandboxed-system-primitives/src/posix.c:2925]: Should check the return value, it error occurs, should os_socket_close(ret)?

File: core/iwasm/libraries/libc-wasi/sandboxed-system-primitives/src/posix.c [hunk 12]
```diff
@@ -2943,3 +3117,88 @@ fd_prestats_destroy(struct fd_prestats *pt)
...
+    cur->next = next;
+    return true;
+}
+
+static bool
+compare_address(const struct addr_pool *addr_pool_entry, const char *addr)
+{
+    /* host order */
+    uint32_t target;
+    uint32_t address = addr_pool_entry->addr;
```
Comment (wenyongh) [inline core/iwasm/libraries/libc-wasi/sandboxed-system-primitives/src/posix.c:3164]: Had better use uint32 but not uint32_t here, we usually use uint32/uint16 like types inside runtime, and use uint32_t/uint16_t like types in header files of core/iwasm/include

File: core/iwasm/libraries/libc-wasi/sandboxed-system-primitives/src/posix.c [hunk 13]
```diff
@@ -2943,3 +3117,88 @@ fd_prestats_destroy(struct fd_prestats *pt)
...
+        return true;
+    }
+
+    if (os_socket_inet_network(addr, &target) != BHT_OK) {
+        return false;
+    }
+
+    uint32_t mask = addr_pool_entry->mask;
+    uint32_t first_address = address & mask;
+    uint32_t last_address = address | (~mask);
```
Comment (wenyongh) [inline core/iwasm/libraries/libc-wasi/sandboxed-system-primitives/src/posix.c:3176]: Same as above

File: core/iwasm/libraries/libc-wasi/sandboxed-system-primitives/src/posix.h [hunk 14]
```diff
@@ -46,6 +46,13 @@ struct argv_environ_values {
     size_t environ_count;
 };
 
+struct addr_pool {
+    struct addr_pool *next;
+    /* addr and mask in host order */
+    uint32_t addr;
+    uint8_t mask;
```
Comment (wenyongh) [inline core/iwasm/libraries/libc-wasi/sandboxed-system-primitives/src/posix.h:53]: Same as above, had better use uint32 and uint8

File: product-mini/platforms/posix/main.c [hunk 15]
```diff
@@ -312,9 +315,34 @@ main(int argc, char *argv[])
                 return print_help();
             }
         }
+        else if (!strncmp(argv[0],
+                          "--addr-pool-file=", strlen("--addr-pool-file="))) {
+            /* --addr-pool-file=/path/to/configuration */
+            /* TODO: parse the configuration file */
+        }
+        else if (!strncmp(argv[0], "--addr-pool=", strlen("--addr-pool="))) {
+            /* --addr-pool=100.200.244.255/30 */
```
Comment (wenyongh) [inline product-mini/platforms/posix/main.c:324]: Had better remove the unused hardcode code

File: product-mini/platforms/posix/main.c [hunk 16]
```diff
@@ -312,9 +315,34 @@ main(int argc, char *argv[])
...
+            char *token = NULL;
+
+            if ('\0' == argv[0][13])
```
Comment (wenyongh) [inline product-mini/platforms/posix/main.c:327]: Should be argv[0][12]

File: core/iwasm/libraries/libc-wasi/sandboxed-system-primitives/include/wasmtime_ssp.h (hunk 5 above)
Comment (lum1n0us) [reply core/iwasm/libraries/libc-wasi/sandboxed-system-primitives/include/wasmtime_ssp.h:597]: Thanks for bringing it. Will run more test under x86-32

File: core/iwasm/libraries/libc-wasi/sandboxed-system-primitives/src/posix.c (hunk 12 above)
Comment (lum1n0us) [reply core/iwasm/libraries/libc-wasi/sandboxed-system-primitives/src/posix.c:3164]: Got it.

File: core/iwasm/common/wasm_runtime_common.c [hunk 17]
```diff
@@ -1969,14 +1969,36 @@ wasm_runtime_set_wasi_args(WASMModuleCommon *module, const char *dir_list[],
                                   argc, -1, -1, -1);
 }
 
+void
+wasm_runtime_set_wasi_addr_pool(wasm_module_t module, const char *addr_pool[],
+                                uint32_t addr_pool_size)
```
Comment (wenyongh) [inline core/iwasm/common/wasm_runtime_common.c:1974]: uint32_t to uint32

File: core/iwasm/common/wasm_runtime_common.h [hunk 18]
```diff
@@ -731,6 +733,9 @@ wasm_runtime_set_wasi_ctx(WASMModuleInstanceCommon *module_inst,
 WASIContext *
 wasm_runtime_get_wasi_ctx(WASMModuleInstanceCommon *module_inst);
 
+WASM_RUNTIME_API_EXTERN void
+wasm_runtime_set_wasi_addr_pool(wasm_module_t module, const char *addr_pool[],
+                                uint32_t addr_pool_size);
```
Comment (wenyongh) [inline core/iwasm/common/wasm_runtime_common.h:738]: uint32_t to uint32

**Pull request merged**